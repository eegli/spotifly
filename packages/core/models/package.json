{
  "main": "dist/spotifly-core-models.cjs.js",
  "module": "dist/spotifly-core-models.esm.js"
}
