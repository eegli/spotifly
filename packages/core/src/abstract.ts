import { AuthProvider } from './request';

export abstract class SpotifyKind {
  protected abstract endpoints: Record<string, Record<string, string | number>>;
  protected request: typeof AuthProvider.prototype.request;

  constructor(provider: InstanceType<typeof AuthProvider>) {
    this.request = provider.request.bind(provider);
  }
}
