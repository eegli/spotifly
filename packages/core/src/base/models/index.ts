export * from './artists';
export * from './tracks/tracks';
