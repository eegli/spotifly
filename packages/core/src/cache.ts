/* eslint-disable @typescript-eslint/no-explicit-any */
/* eslint-disable  @typescript-eslint/no-empty-function */

export enum CacheEntity {
  Artist = 'artist',
  Track = 'track',
}

export function Cacheable<T extends unknown[]>(
  entity: CacheEntity,
  debug = false
) {
  return function cacheMeBaby(
    _: unknown,
    propertyKey: string,
    descriptor: PropertyDescriptor
  ) {
    const log = debug ? console.log : () => {};
    const original = descriptor.value;
    descriptor.value = async function (...args: string[]) {
      if (!Array.isArray(args) || !args.every(arg => typeof arg === 'string')) {
        throw new Error(
          `Decorator applied on uncacheable method ${propertyKey}`
        );
      }
      const { cached, uncached } = cache.filterCached(entity, args);

      if (uncached.length === 0) {
        log('Reading all items from cache');
        return Promise.resolve(cached);
      }

      log(
        `Found ${cached.length}/${args.length} entries in cache, optimizing query...`
      );
      args = uncached;

      const result: T[] = await original.call(this, ...args);

      log(`Putting ${result.length} new item(s) into cache`);
      for (let i = 0; i < result.length; i++) {
        cache.put(entity, [uncached[i], result[i]]);
      }
      log(`Appending ${cached.length} item(s) to result from cache`);
      result.push(...cached);
      return Promise.resolve(result);
    };
  };
}

export type FilteredCache = {
  cached: any[];
  uncached: string[];
};

class Cache {
  #tracks = new Map<string, any>();
  #artists = new Map<string, any>();

  #enabled = true;

  #findBin(entity: CacheEntity) {
    switch (entity) {
      case CacheEntity.Track:
        return this.#tracks;
      case CacheEntity.Artist:
        return this.#artists;
    }
  }

  public disable() {
    this.#enabled = false;
  }

  public put(entity: CacheEntity, item: [string, any]): void {
    const bin = this.#findBin(entity);
    bin.set(item[0], item[1]);
  }

  public filterCached(entity: CacheEntity, ids: string[]): FilteredCache {
    if (!this.#enabled) return { cached: [], uncached: ids };
    const bin = this.#findBin(entity);
    return ids.reduce(
      (acc, curr) => {
        const item = bin.get(curr);
        if (item) {
          acc.cached.push(item);
        } else {
          acc.uncached.push(curr);
        }
        return acc;
      },
      {
        cached: [],
        uncached: [],
      } as FilteredCache
    );
  }
}

export const cache = new Cache();
