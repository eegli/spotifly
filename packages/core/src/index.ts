import { AuthProvider, AuthProviderOptions } from './provider';

import Albums from './models/albums';
import Artists from './models/artists';
import Categories from './models/categories';
import Episodes from './models/episodes';
import future from './models/future';
import Genres from './models/genres';
import Markets from './models/markets';
import Player from './models/player';
import Playlists from './models/playlists';
import Search from './models/search';
import Shows from './models/shows';
import Tracks from './models/tracks';
import Users from './models/users';

export function initialize(authOptions: AuthProviderOptions) {
  const provider = new AuthProvider(authOptions);
  return {
    Albums: Albums(provider),
    Artists: Artists(provider),
    Categories: Categories(provider),
    Episodes: Episodes(provider),
    Genres: Genres(provider),
    Markets: Markets(provider),
    Player: Player(provider),
    Playlists: Playlists(provider),
    Search: Search(provider),
    Shows: Shows(provider),
    Tracks: Tracks(provider),
    Users: Users(provider),
    future: future(provider),
  } as const;
}

export type SpotifyClient = ReturnType<typeof initialize>;

export { isError } from './request';
export type { DataPromise, DataResponse } from './types';
export type { AuthProviderOptions };
