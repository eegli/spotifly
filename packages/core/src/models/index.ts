export * from './artists/artists';
export * from './tracks';
