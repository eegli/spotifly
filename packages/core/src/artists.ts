import { RequestAble, SpotifyKind } from './abstract';
import { MapCollection } from './types';
import { chunkify } from './utils';

export class Artists extends SpotifyKind<SpotifyApi.ArtistObjectFull[]> {
  private readonly MULTIPLE_ARTISTS = '/artists';
  private readonly CHUNK_SIZE = 50;

  private collection: MapCollection<SpotifyApi.ArtistObjectFull> = new Map();
  private artistIds: Set<string> = new Set();

  constructor(opts: RequestAble, id?: string, ...ids: string[]) {
    super(opts);
    if (id) {
      this.add(id, ...ids);
    }
  }

  add(id: string, ...ids: string[]) {
    [id, ...ids].forEach(id => this.artistIds.add(id));
  }

  get items() {
    return Array.from(this.collection.values());
  }

  async get(id: string, ...ids: string[]): Promise<void> {
    for await (const chunk of chunkify([id, ...ids], this.CHUNK_SIZE)) {
      const { data } =
        await this.request.get<SpotifyApi.MultipleArtistsResponse>(
          this.BASE_URL_V1 + this.MULTIPLE_ARTISTS,
          {
            params: {
              ids: chunk.join(','),
            },
          }
        );
      data.artists.forEach(artist => {
        this.collection.set(artist.id, artist);
      });
    }
  }
}
