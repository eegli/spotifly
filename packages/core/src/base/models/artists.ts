import { MapCollection } from '../../types';
import { chunkify } from '../../utils';
import { Collectible, RequestConfig, SpotifyKind } from '../abstract';

export class Artists extends SpotifyKind {
  #artists: MapCollection<SpotifyApi.ArtistObjectFull> = new Map();

  constructor(opts: RequestConfig) {
    super(opts);
  }

  public get collection() {
    return {
      artists: new Collectible(this.#artists),
    };
  }

  async getSeveralArtists(id: string, ...ids: string[]): Promise<this> {
    const chunkSize = this.endpoints.getSeveralArtists.limit;
    for await (const chunk of chunkify([id, ...ids], chunkSize)) {
      const { data } =
        await this.request.get<SpotifyApi.MultipleArtistsResponse>(
          this.endpoints.getSeveralArtists.url,
          {
            params: {
              ids: chunk.join(','),
            },
          }
        );
      data.artists.forEach(artist => {
        this.#artists.set(artist.id, artist);
      });
    }
    return this;
  }
}
