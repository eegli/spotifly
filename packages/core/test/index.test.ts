import { Models } from '../src/models';

describe('Lib exports', () => {
  test('Models', () => {
    expect(Models.GetSeveralTracks).toBeDefined();
  });
});
