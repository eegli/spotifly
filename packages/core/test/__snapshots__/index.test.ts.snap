// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Lib matches snapshot 1`] = `
{
  "Albums": {
    "Album": {
      "get": [Function],
      "getAll": [Function],
      "getSeveral": [Function],
    },
    "NewReleases": {
      "get": [Function],
    },
    "Tracks": {
      "get": [Function],
      "getAll": [Function],
    },
    "UsersSaved": {
      "check": [Function],
      "checkAll": [Function],
      "get": [Function],
      "getAll": [Function],
      "remove": [Function],
      "removeAll": [Function],
      "save": [Function],
      "saveAll": [Function],
    },
  },
  "Artists": {
    "Albums": {
      "get": [Function],
      "getAll": [Function],
    },
    "Artist": {
      "get": [Function],
      "getAll": [Function],
      "getSeveral": [Function],
    },
    "RelatedArtists": {
      "get": [Function],
    },
    "TopTracks": {
      "get": [Function],
    },
  },
  "Categories": {
    "getCategory": [Function],
    "getSeveralCategories": [Function],
  },
  "Episodes": {
    "checkAllUsersSavedEpisodes": [Function],
    "checkUsersSavedEpisodes": [Function],
    "getAllEpisodes": [Function],
    "getAllUsersSavedEpisodes": [Function],
    "getEpisode": [Function],
    "getSeveralEpisodes": [Function],
    "getUsersSavedEpisodes": [Function],
    "removeAllUsersSavedEpisodes": [Function],
    "removeUsersSavedEpisodes": [Function],
    "saveAllEpisodesForUser": [Function],
    "saveEpisodesForUser": [Function],
  },
  "Genres": {
    "getAvailableGenreSeeds": [Function],
  },
  "Markets": {
    "getAvailableMarkets": [Function],
  },
  "Player": {
    "getAvailableDevices": [Function],
    "getPlaybackState": [Function],
    "transferPlayback": [Function],
  },
  "Playlists": {
    "addPlaylistItems": [Function],
    "changePlaylist": [Function],
    "createPlaylist": [Function],
    "getAllCurrentUsersPlaylists": [Function],
    "getAllPlaylistItems": [Function],
    "getAllUsersPlaylists": [Function],
    "getCategoryPlaylists": [Function],
    "getCurrentUsersPlaylists": [Function],
    "getFeaturedPlaylists": [Function],
    "getPlaylist": [Function],
    "getPlaylistCoverImage": [Function],
    "getPlaylistItems": [Function],
    "getUsersPlaylists": [Function],
    "removePlaylistItems": [Function],
    "reorderPlaylistItems": [Function],
  },
  "Search": {
    "search": [Function],
  },
  "Shows": {
    "Episodes": {
      "get": [Function],
      "getAll": [Function],
    },
    "Show": {
      "get": [Function],
      "getAll": [Function],
      "getSeveral": [Function],
    },
    "UsersSaved": {
      "check": [Function],
      "checkAll": [Function],
      "get": [Function],
      "getAll": [Function],
      "remove": [Function],
      "removeAll": [Function],
      "save": [Function],
      "saveAll": [Function],
    },
  },
  "Tracks": {
    "checkAllUsersSavedTracks": [Function],
    "checkUsersSavedTracks": [Function],
    "getAllAudioFeatures": [Function],
    "getAllTracks": [Function],
    "getAllUsersSavedTracks": [Function],
    "getAudioAnalysis": [Function],
    "getAudioFeatures": [Function],
    "getRecommendations": [Function],
    "getSeveralAudioFeatures": [Function],
    "getSeveralTracks": [Function],
    "getTrack": [Function],
    "getUsersSavedTracks": [Function],
    "removeAllUsersTracksForUser": [Function],
    "removeUsersSavedTracks": [Function],
    "saveAllTracksForUser": [Function],
    "saveTracksForUser": [Function],
  },
  "Users": {
    "checkFollowsAllArtists": [Function],
    "checkFollowsAllUsers": [Function],
    "checkFollowsArtists": [Function],
    "checkFollowsUsers": [Function],
    "checkUsersFollowPlaylist": [Function],
    "followAllArtists": [Function],
    "followAllUsers": [Function],
    "followArtists": [Function],
    "followPlaylist": [Function],
    "followUsers": [Function],
    "getAllUsersTopArtists": [Function],
    "getAllUsersTopTracks": [Function],
    "getCurrentUsersProfile": [Function],
    "getUsersFollowedArtists": [Function],
    "getUsersProfile": [Function],
    "getUsersTopArtists": [Function],
    "getUsersTopTracks": [Function],
    "unfollowAllArtists": [Function],
    "unfollowAllUsers": [Function],
    "unfollowArtists": [Function],
    "unfollowPlaylist": [Function],
    "unfollowUsers": [Function],
  },
}
`;
